src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/PlatformLayouts.java

```diff
@@ -52,47 +50,56 @@ public static MemoryLayout asVarArg(MemoryLayout ml) {
     }
 
-    private static CValueLayout ofInt(ByteOrder order, long bitSize) {
-        return new CValueLayout(CValueLayout.Kind.INT, order, bitSize, bitSize, Map.of());
+    private static ValueLayout ofInt(ByteOrder order, long bitSize) {
+        return MemoryLayout.ofValueBits(bitSize, order)
+                .withAttribute(CLinker.TypeKind.ATTR_NAME, CLinker.TypeKind.INT);
     }
 
-    private static CValueLayout ofLong(ByteOrder order, long bitSize) {
-        return new CValueLayout(CValueLayout.Kind.LONG, order, bitSize, bitSize, Map.of());
+    private static ValueLayout ofLong(ByteOrder order, long bitSize) {
+        return MemoryLayout.ofValueBits(bitSize, order)
+                .withAttribute(CLinker.TypeKind.ATTR_NAME, CLinker.TypeKind.LONG);
     }
 
-    private static CValueLayout ofLongLong(ByteOrder order, long bitSize) {
-        return new CValueLayout(CValueLayout.Kind.LONGLONG, order, bitSize, bitSize, Map.of());
+    private static ValueLayout ofLongLong(ByteOrder order, long bitSize) {
+        return MemoryLayout.ofValueBits(bitSize, order)
+                .withAttribute(CLinker.TypeKind.ATTR_NAME, CLinker.TypeKind.LONGLONG);
     }
 
-    private static CValueLayout ofFloat(ByteOrder order, long bitSize) {
-        return new CValueLayout(CValueLayout.Kind.FLOAT, order, bitSize, bitSize, Map.of());
+    private static ValueLayout ofFloat(ByteOrder order, long bitSize) {
+        return MemoryLayout.ofValueBits(bitSize, order)
+                .withAttribute(CLinker.TypeKind.ATTR_NAME, CLinker.TypeKind.FLOAT);
     }
 
-    private static CValueLayout ofDouble(ByteOrder order, long bitSize) {
-        return new CValueLayout(CValueLayout.Kind.DOUBLE, order, bitSize, bitSize, Map.of());
+    private static ValueLayout ofDouble(ByteOrder order, long bitSize) {
+        return MemoryLayout.ofValueBits(bitSize, order)
+                .withAttribute(CLinker.TypeKind.ATTR_NAME, CLinker.TypeKind.DOUBLE);
     }
 
-    private static CValueLayout ofLongDouble(ByteOrder order, long bitSize) {
-        return new CValueLayout(CValueLayout.Kind.LONGDOUBLE, order, bitSize, bitSize, Map.of());
+    private static ValueLayout ofLongDouble(ByteOrder order, long bitSize) {
+        return MemoryLayout.ofValueBits(bitSize, order)
+                .withAttribute(CLinker.TypeKind.ATTR_NAME, CLinker.TypeKind.LONGDOUBLE);
     }
 
     /**
-     * Creates a new CValueLayout with the {@code POINTER} kind
+     * Creates a new ValueLayout with the {@code POINTER} kind
      *
      * @param order the byte order of the layout
      * @param bitSize the size, in bits, of the layout
-     * @return the newly created CValueLayout
+     * @return the newly created ValueLayout
      */
-    public static CValueLayout ofPointer(ByteOrder order, long bitSize) {
-        return new CValueLayout(CValueLayout.Kind.POINTER, order, bitSize, bitSize, Map.of());
+    public static ValueLayout ofPointer(ByteOrder order, long bitSize) {
```

Comment:
Looks like this can be private? (and in that case I think it's okay to drop the javadoc as well, since none of the other factories have it either.)